### DISCUSSION TITLE
FetchError: res.text is not a function

### DISCUSSION THREAD

**Discussion by RajisteB** [supabase/supabase#13085] [Questions]
I've been looking all over online to see if this has come up before? No avail in discord either. I'm using Node js with the @supabase/supabase-js module and I'm trying to insert a single record and keep getting this error. 

```
const { error } = await supabase
    .from("cryptocurrency")
    .insert({symbol: 'BTC', name: 'Bitcoin'...});


// Error

  error: {
    message: 'FetchError: res.text is not a function',
    details: '',
    hint: '',
    code: ''
  }
```

**Comment by Goli4thus** [comment]
I stumbled over a similar error just now and this really still was one of the only result I could find (sadly no answers yet).

## The error message
In my case, I had this error:
```
 message: 'TypeError: res.text is not a function',
 details: 'TypeError: res.text is not a function\n' +
   '    at /....<project_path>....../node_modules/@supabase/postgrest-js/src/PostgrestBuilder.ts:84:34\n' +
```

Looking into that file, it sure had lines like this:
```
const body = await res.text()
```


## The reason for why I even got this error in the first place:
I was implementing some automated tests using `jest`. While doing so, I needed to mock `fetch`, cause the to be tested code was making some API calls. 

I had decided to try mocking it manually. The intended mocking eventually worked so to speak. But based on above error, I kind
of suspected that this mocked `fetch` version was getting in the way of `supabase-js` (or `postgres-js` for that matter).

## The eventual solution
It was about extending the mocked `fetch` version to the point of it transparently forwarding all "not to be mocked fetch
requests".
The gist of my mock code that got rid of the above error is this:

```
// globals.d.ts
declare module global {
    var fetch: {};
}

// whatever.test.ts
declare const fetch: CallableFunction
const unmockedFetch = global.fetch as CallableFunction;

beforeAll(() => {
    global.fetch = jest.fn(async (a_url: string, a_options: { body: any }) => {
        if (a_url.startsWith("https://<to-be-mocked-api-url>")) {
            return Promise.resolve({
                ok: true,
                json: async (url = a_url, options = a_options) => {
                    console.log(`from fetch mock (url): ${url}`);

                    const data = { answer: "some test" };

                    return Promise.resolve(data);
                }
            })
        } else {
            // This is the "transparent forwarding" part.
            return unmockedFetch(a_url, a_options);
        }
    }) as jest.Mock;
});
afterAll(() => {
    global.fetch = unmockedFetch;
});
```

## Closing thoughts
There sure were some hoops to jump through re typescript to make it work without getting various type errors.

Potentially it would have been solved quicker by using something like this: https://www.npmjs.com/package/jest-fetch-mock
Then again, one likely ends up learning more about mocking in jest when doing things like this.
